RFD.md

```diff
@@ -0,0 +1,217 @@
+
+* We're going to create a new clean filesystem by extracting an Alpine Root FS image per command, which will use 5Mb of disk space per container. In a production environment we would probably rely on some kind of layered filesystem like overlayfs to save on disk resources.
+
+#### Availability and resource limits
+
+In a multi-tenant environment like the system in question, we need to ensure proper resource limits to reduce the potential impact of noisy or abusive clients on the underlying host OS.
+
+We're going to use linux Cgroups to apply the following limits on each command:
+
+* Memory usage – a static per-command limit will be applied on each container.
+
+* CPU usage - each command will be limited to a small constant amount of CPU resources, preventing it from negative affecting the host OS and other commands by consuming too much CPU.
+
+* Disk IO - each command will be limited by the amount of disk IO it could perform to limit the effects from a single command's high IO load on the host OS and other commands.
+
+##### Scope limits
+
+All limits will be statically hardcoded within the server, while in a production system we'd probably have some kind of configuration file for default limits and would expand the API to allow users to specify their own limits for each command (like Docker does).
+
+In a production environment, we may want to introduce per-user resource limits, API call rate limits, etc to further protect the system from abuse.
+
+### API
+
+The following API methods will be available on the server:
+
+* `Status` - returns some basic status information for the server, including the list of running and finished commands (for non-admin users only the processes owned by the user would be visible).
+* `StartCommand` - starts a command on the server and returns a unique `command_id` value (a UUID string) used to manage the command in subsequent API calls.
+* `StopCommand` - stops a given command (identified by a `command_id`).
+* `CommandStatus` - returns current status information for a given command (identified by a `command_id`).
+* `CommandOutput` - a streaming API for receiving console output (combined stdout+stderr) from a given command (identified by a `command_id`). If requested, the stream will continue until the command has finished (tail mode).
+
+The detailed GRPC definition for the proposed API can be found within the [remote_exec.proto](remote_exec/remote_exec.proto) file.
+
+### Library design
+
+Most of concerns around containerization will be isolated within the `container_exec` library, allowing us to have a very simple GRPC server around the library. This should make it easier to test the containerization code and contain most of the complexity (dealing with Linux APIs, process management, logs management, concurrency concerns, etc) the library.
```

Comment:
Just to be clear, you're writing `container_exec` and not re-using something right?

Reply:
Yes :-) Unless reusing something of that magnitude (like some pieces from Moby or containerd) is actually an option, of course ;-)

Reply:
:+1: to creating own library.